### DISCUSSION TITLE
Refactoring 1: State struct

### DISCUSSION THREAD

**Discussion by TornaxO7** [leftwm/leftwm#885] [Ideas]
So I'd like to start with a proposal for the `State` struct:

```rust
pub struct State { 
    /// screens is known in each workspace
    // pub screens: Vec<Screen>,
    /// windows can be collected through the tags
    // pub windows: Vec<Window>,

    pub workspaces: Vec<Workspace>,

    /// this should be stored in each workspace/tag
    // pub focus_manager: FocusManager, 
    pub mode: Mode,
    /// this can be read from the config struct
    // pub layouts: Vec<Layout>,
    /// this can be read fvrom the foncig struct
    //pub scratchpads: Vec<ScratchPad>,
    /// is it not posible to handle the scratchpads as windows?
    pub active_scratchpads: HashMap<ScratchPadName, VecDeque<ChildID>>,
    pub actions: VecDeque<DisplayAction>,
    pub tags: Tags, // List of all known tags.
    /// if I'm right, this is also in the config
    // pub mousekey: Vec<String>,
    /// those values can be read from the config
    // pub max_window_width: Option<Size>,
    // pub default_width: i32,
    // pub default_height: i32,
    // pub disable_tile_drag: bool,
    // pub insert_behavior: InsertBehavior,

    /// this can be skipped since the state of the config shouldn't change
    /// during a session
    #[serde(skip)]
    config: Box<dyn Config>,
}
```

to sum it up, it would look like this:
```rust
pub struct State {
    pub workspaces: Vec<Workspace>,
    pub tags: Tags,
    pub mode: Mode,
    pub active_scratchpads: HashMap<ScratchPadName, VecDeque<ChildID>>,
    pub actions: VecDeque<DisplayAction>,

    #[serde(skip)]
    config: Box<dyn Config>,
}
```
Any thoughts about this?

**Comment by VuiMuich** [comment]
I am not sure what gain you expect from this refactor. 

The state is in my understanding a performance optimised communication hub between all the different modules. 
If you replace all these fields with methods that fetch it from the other places, this might have a significant impact on performance especially for tasks that need to collect the whole state, like `leftwm-state` does on every update to a single parameter in order to print it to stdout. Storing and reloading the state for a soft reload is not so much of an issue right now, as this is a process that is not completely butter smooth anyways. But lag in output of `leftwm-state` would be a bad thing.

I am happily proven wrong on this though :smiley:
Plus some points are really debatable what their purpose in the state is, like the `mousekey` and I am open to ideas, how to refactor the `active_scratchpads` as they indeed would feel more naturally at home within something window related.

On the 'windows can be collected through tags': are you considering that tags are not holding a list of windows it owns, but each window holds a tag it belongs to. So who would hold all the windows if not the state?

**Reply by TornaxO7** [reply]
> I am not sure what gain you expect from this refactor.

I mainly expect to reduce the complexity of information for each struct with the minimal information they actually need.

> The state is in my understanding a performance optimised communication hub between all the different modules.

oh. I thought, due to the name, that the state should hold the main-information of the current leftwm-session data. 

> Plus some points are really debatable what their purpose in the state is, like the mousekey and I am open to ideas,

yes, that's why the state should hold the config in my opinion (which can be ignored by the serealizer) to avoid redundant information-storage.

>  So who would hold all the windows if not the state?

I'm saying that a tag should hold a list of windows.

> If you replace all these fields with methods that fetch it from the other places, this might have a significant impact on performance especially for tasks that need to collect the whole state

I don't really think that this will take more significantly more time. Let's take the window-collection as an example. In my opinion it should look like this:
```rust
pub struct State {
     tags: Tags,
}

impl State {
     // ...

    pub fn get_windows(&self) -> Vec<Windows> {
        let mut windows = Vec::new();
        // this needs O(a)
        for tag in self.tags.all() {
             // cloning takes O(1) since the window contains only atrributes with one value except `states` which doesn't hold the same value twice I think
            // and `extending` should take O(k_n) time, where k_n is the amount of windows in tag 
            windows.extend(tag.windows.clone());
        } 
        // so in summary we have a + k_1 + k_2 + ... + k_n where the sum of k_n equals n (the amount of windows) since it's like going through all the windows once and we can ignore `a` since it's a constant so in summary we are having O(n) here.
       // Mostly for all operation on the windows, we are iterating over it as well, so it will take O(n) here as well. In summary this means
       // we are taking n + n \in O(n).

       windows
    }
}
```
